### PR TITLE
Add response-type=token to login/oauth/authorize for single page apps

### PR DESCRIPTION
**Pull request by cadecairos**
#356 

R?

It doesn't say that a client shouldn't be able to use it in any way - should we add that in? how should we do it? a new column?

### CODE REVIEW

**Comment by cadecairos** [comment]
@jbuck Added in allowed_responses to verify the client can indeed use the provided response_type

File: lib/queries.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ module.exports = {
       "access_tokens WHERE access_token = $1;"
   },
   create: {
-    client: "INSERT INTO clients VALUES($1, $2, $3, $4);",
+    client: "INSERT INTO clients VALUES($1, $2, $3, $4, $5);",
```
Comment (jbuck) [inline lib/queries.js]: Could you name these columns explicitly? I worry about the case where the db column order isn't the same as what this code is expecting. If you wanna fix up the two lines below this too that'd be lovely :)

File: web/server.js [hunk 2]
```diff
@@ -163,6 +163,19 @@ module.exports = function(options) {
       method: 'GET',
       path: '/login/oauth/authorize',
       config: {
+        validate: {
```
Comment (jbuck) [inline web/server.js]: mmmmmm tasty validations

File: web/server.js [hunk 3]
```diff
@@ -163,6 +163,19 @@ module.exports = function(options) {
...
+          query: {
+            client_id: Joi.string().required(),
+            response_type: Joi.string().valid('code', 'token'),
+            scopes: Joi.string().required(),
+            state: Joi.string().when('response_type', {
```
Comment (jbuck) [inline web/server.js]: Let's make state a require parameter, to be consistent

File: web/server.js [hunk 4]
```diff
@@ -163,6 +163,19 @@ module.exports = function(options) {
...
+              is: 'code',
+              then: Joi.required(),
+              otherwise: Joi.forbidden()
+            }),
+            action: Joi.string().optional().valid('signup')
```
Comment (jbuck) [inline web/server.js]: I think this should be `.default('signin')` right?

File: web/server.js [hunk 5]
```diff
@@ -190,36 +203,49 @@ module.exports = function(options) {
...
-        redirectObj.query.client_id = request.query.client_id;
-        var redirectUri = url.format(redirectObj);
-
-        reply.redirect(redirectUri);
+        var expiresAt = new Date(Date.now() + 60 * 1000).toISOString();
+        var redirectObj = url.parse(client.redirect_uri, true);
+
+        scopes = scopes.split(' ');
+
+        oauthDb.generateAuthCode(
```
Comment (jbuck) [inline web/server.js]: Lets move this generateAuthCode function back into a pre method, and put the generateToken function in a pre method as well. We can have them conditionally run based on the response_type, and also return an error if it's not allowed. Then the handler can just be output formatting!

**Comment by jbuck** [comment]
r-, needs some code fixings

File: docs/oauth.md [hunk 6]
```diff
@@ -258,6 +258,51 @@ To access the API using your access token, set the `Authorization` header for ea
...
+  <tr>
+    <td>response_type</td>
+    <td>The response_type must be set to "token".</td>
+  </tr>
+  <tr>
+    <td>action</td>
+    <td>Optional. If provided, should be <code>signup</code> or <code>signin</code>. Send to improve the user experience, based on whether the user clicked on a Sign In or Sign Up button. <code>signin</code> is the default.</td>
+  </tr>
+  <tr>
+    <td>scopes</td>
```
Comment (jbuck) [inline docs/oauth.md]: Can you add the `state` parameter to the docs here?

**Comment by jbuck** [comment]
@cadecairos r+

**Pull request merged**